### PR TITLE
Fix pytorch memory curve estimation for rmm backed allocator 

### PR DESCRIPTION
**Pull request by VibhuJawa**
This PR fixes https://github.com/rapidsai/crossfit/issues/74 by working around  https://github.com/pytorch/pytorch/issues/133281 . 

Also fix: https://github.com/rapidsai/crossfit/issues/92

### CODE REVIEW

File: crossfit/utils/torch_utils.py [hunk 1]
```diff
@@ -99,3 +100,44 @@ def cleanup_torch_cache() -> None:
...
+    If RMM is being used as the allocator, it enables RMM statistics and pushes a new
+    statistics context. If the default PyTorch allocator is being used, it resets the
+    peak memory stats for CUDA.
+
+    Returns:
+        None
+    """
+    # TODO: This is hacky, we need to check if the allocator is rmm
+    #  and then reset the peak memory stats
+    if torch.cuda.memory.get_allocator_backend() == "pluggable":
```
Comment (VibhuJawa) [inline crossfit/utils/torch_utils.py:122]: @bdice , Curious if you know of a there a better way to check this, this feels very hacky.

File: tests/op/test_fit_memory_estimate.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+torch = pytest.importorskip("torch")
+rmm_torch_allocator = pytest.importorskip(
+    "rmm.allocators.torch", reason="rmm_torch_allocator is not available."
+).rmm_torch_allocator
+
+MODEL_NAME = "microsoft/deberta-v3-base"
+
+# Have to do it globally
+# TODO: Ask for better way
+torch.cuda.memory.change_current_allocator(rmm_torch_allocator)
```
Comment (VibhuJawa) [inline tests/op/test_fit_memory_estimate.py:16]: Curious if we can avoid doing this globally.

File: crossfit/backend/torch/hf/memory_curve_utils.py [hunk 3]
```diff
@@ -37,7 +41,7 @@ def fit_memory_estimate_curve(
 ) -> LinearRegression:
     print(f"Fitting memory estimate curve for model: {path_or_name}")
 
-    device = next(model.parameters()).device
+    device = "cuda"
```
Comment (VibhuJawa) [inline crossfit/backend/torch/hf/memory_curve_utils.py:44]: This closes: https://github.com/rapidsai/crossfit/issues/92

File: crossfit/utils/torch_utils.py (hunk 1 above)
Comment (sarahyurick) [reply crossfit/utils/torch_utils.py:121]: I haven't used any of these before but would something like [`_get_current_allocator`](https://github.com/pytorch/pytorch/blob/22e19bd2d70409c2908edf0b0d00abb9209e3aaa/torch/cuda/memory.py#L972C16-L972C27) work? Or maybe check if RMM is already initialized with [`is_initialized`](https://github.com/rapidsai/rmm/blob/6489bb7df63a3784b4a94067e3a8fa8917523ab7/python/rmm/rmm/rmm.py#L94)?

Comment (VibhuJawa) [reply crossfit/utils/torch_utils.py:121]: I tried,  but the type is still `torch.cuda`, so unsure how to check  using that

```
In [12]: type(torch.cuda.memory._get_current_allocator())
Out[12]: torch.cuda.memory._CUDAAllocator
```

Comment (bdice) [reply crossfit/utils/torch_utils.py:121]: It doesn't look like Torch tracks anything except the allocate and deallocate function pointers. I don't think that is enough information to know what backend was used. You would probably need to track this in some global state when enabling the RMM allocator.

Comment (VibhuJawa) [reply crossfit/utils/torch_utils.py:121]: Thanks Bradley, let me think of a better way to do this. 

Unfortunately a global state may not always work. I am pushing some dask level changes (https://github.com/rapidsai/dask-cuda/pull/1392), I can use that but dont think i can always force users to use this method .

Comment (VibhuJawa) [reply crossfit/utils/torch_utils.py:121]: We dont have a good way to handle this today. Will follow up on Pytorch, resolving for now

**Review by praateekmahajan — APPROVED** [review]

File: crossfit/backend/torch/hf/memory_curve_utils.py (hunk 3 above)
Comment (praateekmahajan) [inline crossfit/backend/torch/hf/memory_curve_utils.py:44]: Would it be safer to do
```
try:
    device = next(model.parameters()).device
except AttributeError as e: # unsure if AttributeError is raised
    device = "cuda"
```

Comment (VibhuJawa) [reply crossfit/backend/torch/hf/memory_curve_utils.py:44]: Actually i want all estimation to happen on `cuda` and if it is not then complain loudly.

**Review by praateekmahajan — APPROVED** [review]

**Pull request merged**